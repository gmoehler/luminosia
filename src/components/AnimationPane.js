import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { getChannelExportData } from '../actions/generalActions';
import { samplesToRad } from '../utils/conversions';
import Slider from '@material-ui/lab/Slider';

const minRotationSpeed = 0.1;
const maxRotationSpeed = 10;


const AnimationPaneWrapper = styled.div`
	width:  calc(95vw - ${props => props.drawerWidth}px);
	background: black;
	display: flex;
	flex-orientation: row;
`;

const AnimationControl = styled.div`
	width:  96px;
	background: darkgrey;
`;

const AnimationCanvas = styled.canvas`
`;

export default class AnimationPane extends PureComponent {
	constructor(props) {
		super(props);
		this.prevRad = 0;
		this.prevTime = 0;
		this.innerRadius = 2;
		this.margin = 10;
		this.state = {
			rotationSpeed: 2, // rotations per second
		}
	}


	drawArc(cc, arcIdx, color, radius, fromRad, toRad) {

		const {resolution} = this.props;

		const largestRadius = resolution * (30 + this.innerRadius);
		const centerX = this.margin + largestRadius + arcIdx * (this.margin + 2 * largestRadius);
		const centerY = this.margin + largestRadius;

		cc.beginPath();
		cc.arc(centerX , centerY, resolution*radius, fromRad, toRad, false);
		cc.strokeStyle = color;
		cc.lineWidth = 2;
		cc.stroke();
	}

	draw() {

		const {progress, sampleRate, selectedChannels } = this.props;

		if (progress && progress > 0) {
			// get the part of the image that happened during the last update interval
			const expData = getChannelExportData(this.prevTime, progress, sampleRate);

			const canvas = document.getElementById("animationPaneCanvas");
			// canvas.height = 2* (2*this.innerRadius + 60) + 20;
			const cc = canvas.getContext('2d');

			const oneSampleRad  = samplesToRad(1, sampleRate, this.state.rotationSpeed);
			const d = expData.data;
			const numArcs = selectedChannels.length;

			// console.log(`Num arcs: ${numArcs} (${expData.height} x ${expData.width})`)

				for (let w = 0; w < expData.width; w++) { //left to right (i.e. time)
					const toRad = this.prevRad + oneSampleRad;
					for (let i=0; i < 30; i++) { // top to bottom
						for (let arcIdx=0; arcIdx < numArcs; arcIdx++) { // pois
							const row = 30 * arcIdx + i;
							const startIdx = row * expData.width + w;
							// console.log(`animate: ${row} ${startIdx}`)
							const dataIdx = 4 * startIdx;
							const color = `rgba(${d[dataIdx]},${d[dataIdx+1]},${d[dataIdx+2]},255)`;
							this.drawArc(cc, arcIdx, color, i+this.innerRadius, this.prevRad, toRad)
						}
					}
					this.prevRad = toRad;
			}
			this.prevTime = progress;
	}
}
	
	componentDidMount() {
    this.draw();
  }

  componentDidUpdate() {
    this.draw();
  }
  
  handleChange = (ev, val) => {
  	const rotationSpeed = 10 / (101-val); // 0.1 ... 10
  	this.setState({rotationSpeed});
  };

  render() {
  	
  	const { rotationSpeed } = this.state;

		const {drawerWidth, selectedChannels, resolution} = this.props;
    return (
			<AnimationPaneWrapper
					drawerWidth = {drawerWidth}>
					<AnimationControl>
						<Slider 
							value={rotationSpeed}
							onChange={this.handleChange}
							vertical
						/>
					</AnimationControl>
					<AnimationCanvas 
						id = "animationPaneCanvas" 
						height = {resolution * 80}
						width = {selectedChannels.length * resolution * 80}/>
        </AnimationPaneWrapper>
        
    )
	}
}

AnimationPane.propTypes = {
	progress: PropTypes.number,
	sampleRate: PropTypes.number.isRequired,
}