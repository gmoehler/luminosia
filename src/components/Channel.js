import React, { Component } from 'react';
import styled, { withTheme } from 'styled-components';
import { getMouseEventPosition } from '../utils/eventUtils';

const MAX_CANVAS_WIDTH = 1000;

const Progress = styled.div`
  position: absolute;
  background: ${props => props.theme.waveProgressColor};
  width: ${props => props.progress + props.offset}px;
  height: ${props => props.waveHeight}px;
  border-right: 1px solid ${props => props.theme.waveProgressBorderColor};
`;

const Cursor = styled.div`
  position: absolute;
  background: ${props => props.theme.cursorColor};
  width: 1px;
  left: ${props => props.offset + props.cursorPos}px;
  height: ${props => props.waveHeight}px;
`;

const Marker = styled.div`
  position: absolute;
  background: ${props => props.markerColor || props.theme.markerColor};
  width: 1px;
  left: ${props => props.offset + props.markerPos}px;
  height: ${props => props.waveHeight}px;
`;

const Selection = styled.div`
  position: absolute;
  left: ${props => props.offset + props.selection.from}px;
  background: ${props => props.theme.selectionColor};
  width: ${props => props.selection.to - props.selection.from}px;
  height: ${props => props.waveHeight}px;
`;

const Waveform = styled.canvas`
  float: left;
  position: relative;
  margin: 0;
  padding: 0;
  width: ${props => props.cssWidth}px;
  height: ${props => props.waveHeight}px;
`;

const WaveformCanvases = styled.div`
  float: left;
  position: relative;
  left: ${props => props.offset}px;
  background: ${props => props.theme.waveFillColor};
`;

// need position:relative so children will respect parent margin/padding
const ChannelWrapper = styled.div`
  position: relative; 
  margin: 0;
  padding: 0;
  background: ${props => props.theme.waveOutlineColor};
  width: ${props => props.cssWidth + 2}px;
  height: ${props => props.waveHeight}px;
  border: 1px solid ${props => props.borderColor}; 
`;

class Channel extends Component {
  constructor(props) {
    super(props);
    this.canvases = [];
  }

  componentDidMount() {
    this.draw();
  }

  componentDidUpdate() {
    this.draw();
  }

  draw() {
    const {peaks, bits, /* length,*/ waveHeight, theme, scale} = this.props;

    let offset = 0;
    for (let i = 0; i < this.canvases.length; i++) {
      const canvas = this.canvases[i];
      if (!canvas) {
        break; // TODO: find out how to reset canvases on new render
      }
      const cc = canvas.getContext('2d');
      const h2 = waveHeight / 2;
      const maxValue = 2 ** (bits - 1);

      cc.clearRect(0, 0, canvas.width, canvas.height);
      cc.fillStyle = theme.waveOutlineColor;
      cc.scale(scale, scale);

      const peakSegmentLength = canvas.width / scale;
      for (let i = 0; i < peakSegmentLength; i += 1) {
        const minPeak = peaks[(i + offset) * 2] / maxValue;
        const maxPeak = peaks[((i + offset) * 2) + 1] / maxValue;

        const min = Math.abs(minPeak * h2);
        const max = Math.abs(maxPeak * h2);

        // draw max
        cc.fillRect(i, 0, 1, h2 - max);
        // draw min
        cc.fillRect(i, h2 + min, 1, h2 - min);
      }

      offset += MAX_CANVAS_WIDTH;
    }
  }

  handleMouseEvent = (e, eventName) => {
    if (this.props.handleMouseEvent) {
      const pos = getMouseEventPosition(e, "ChannelWrapper");
      const eventInfo = {
      	...pos, // x pos, channelId, partId
      	timestamp: e.timeStamp,
      	// no drag source path
      }
      this.props.handleMouseEvent(eventName, eventInfo);

      return;
    }
  }

  createCanvasRef(i) {
    return (canvas) => {
      this.canvases[i] = canvas
    }
  }

  render() {
    const {length, waveHeight, scale, progress, cursorPos, 
      selection, markers, theme, offset, selected} = this.props;

    let totalWidth = length;
    let waveformCount = 0;
    const waveforms = [];
    while (totalWidth > 0) {
      const currentWidth = Math.min(totalWidth, MAX_CANVAS_WIDTH);
      const waveform = <Waveform key={ `${length}-${waveformCount}` } cssWidth={ currentWidth } width={ currentWidth * scale } height={ waveHeight * scale } waveHeight={ waveHeight } ref={ this.createCanvasRef(waveformCount) }
      />

      waveforms.push(waveform);
      totalWidth -= currentWidth;
      waveformCount += 1;
    }

    const progressElem = progress ? 
      <Progress className='Progress' progress={ progress } theme={ theme } waveHeight={ waveHeight } offset={offset}/>
      : null;

    const selectionElem = selection && selection.from && selection.to ? 
      <Selection className='Selection' selection={ selection } theme={ theme } waveHeight={ waveHeight } offset={offset}/>
      : null;

    const cursorElem = cursorPos ? 
      <Cursor className='Cursor' cursorPos={ cursorPos } theme={ theme } waveHeight={ waveHeight } offset={offset}/>
      : null;
      
    const markerElems = markers && Array.isArray(markers) ?
      markers.map((marker) => {
        let color = theme.markerColor;
        if ( marker.type  === "insert" || marker.id  === "insert" ) {
          color = theme.insertMarkerColor;
        } else if ( marker.type  === "selected" ) {
          color = theme.selectedMarkerColor;
        }
        return <Marker 
          className='Marker' 
          key={marker.id} 
          markerPos= { marker.pos } 
          markerColor={color} 
          theme={ theme } 
          waveHeight={ waveHeight } 
          offset={offset}/>
      }
      ) : null;

    const borderColor = selected ? theme.borderColorSelected : theme.borderColor;

    return (
      <ChannelWrapper 
        className='ChannelWrapper' 
        onMouseDown={ (e) => this.handleMouseEvent(e, "mouseDown") } 
        onMouseUp={ (e) => this.handleMouseEvent(e, "mouseUp") } 
        onMouseMove={ (e) => this.handleMouseEvent(e, "mouseMove") } 
        onMouseLeave={ (e) => this.handleMouseEvent(e, "mouseLeave") }

        cssWidth={ length } 
        theme={ theme } 
        waveHeight={ waveHeight }

        borderColor={borderColor}>

        <WaveformCanvases className='WaveformCanvases' theme={ theme } offset={offset} >
          { waveforms }
        </WaveformCanvases>
        
        { progressElem }
        { selectionElem }
        { cursorElem }
        { markerElems }
        
      </ChannelWrapper>
      );
  }
}

Channel.defaultProps = {
  theme: {
    // color of the waveform outline
    waveOutlineColor: '#282c34',
    waveFillColor: '#05a0cd',
    waveProgressColor: 'transparent', //'rgb(255,120,0)',
    waveProgressBorderColor: 'rgb(255,255,255)',
    cursorColor: 'red',
    markerColor: 'rgba(255,255, 0, 0.5)', // transparent yellow
    insertMarkerColor: 'rgba(255,165, 0, 0.5)', // transparent orange
    selectedMarkerColor: 'rgba(255,165, 0, 1)', // orange
    selectionColor: 'rgba(0,0,255,0.5)',
    borderColorSelected: '#3f51b5',
    borderColor: 'darkgrey',
  },
  // checking `window.devicePixelRatio` when drawing to canvas.
  scale: 1,
  offset: 0,
  peaks: [],
  length: 0,
  bits: 0,
  // height in CSS pixels of each canvas element a waveform is on.
  waveHeight: 90,
  // all x pixel values are from 0 regardless of offset
  // width in CSS pixels of the progress on the channel  (not drawn on null)
  progress: null,
  // position of the cursor in CSS pixels from the left of channel (not drawn on null)
  cursorPos: null,
  // position of the selection in CSS pixels from the left of channel (not drawn on null)
  selection: null,
  // positions of the markers in CSS pixels from the left of channel (null: do not draw)
  markers: [],
};

export default withTheme(Channel);
