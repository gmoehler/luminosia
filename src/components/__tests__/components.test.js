import React from "react";
import renderer from "react-test-renderer";
import AnimationPane from "../AnimationPane.js";
import Channel from "../AudioChannel";
import { imageChannel1, imageChannel2 } from "../../__fixtures__/entity.fixtures";

describe("simple component tests", () => {
  test("AnimationPane component", () => {

    const props = {
      sampleRate: 1000,
      resolution: 100,
      activeChannels: [
        imageChannel1.channelId, imageChannel2.channelId],
    };

    const component = renderer.create(
      <AnimationPane { ...props } />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Channel component", () => {

    const props = {
      channelId: imageChannel1.channelId,
      selection: {
        from: 0,
        to: 22
      }
    };

    const component = renderer.create(
      <Channel { ...props } />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });


});