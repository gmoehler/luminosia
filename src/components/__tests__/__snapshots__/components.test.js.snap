// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple component tests AnimationPane component 1`] = `
<div
  className="sc-ifAKCX kuDDyM"
>
  <div
    className="sc-EHOje kRjwTe"
  >
    <p
      className="MuiTypography-root sc-bZQynM ebBIAI MuiTypography-body1"
    >
       
      2.0
    </p>
    <div
      className="MuiSlider-root sc-bxivhb dKSUMk MuiSlider-vertical"
      onClick={[Function]}
      onMouseDown={[Function]}
      onTouchMove={[Function]}
      onTouchStartCapture={[Function]}
      style={
        Object {
          "width": 0,
        }
      }
    >
      <div
        className="MuiSlider-container MuiSlider-vertical"
      >
        <div
          className="MuiSlider-track MuiSlider-trackBefore MuiSlider-vertical"
          style={
            Object {
              "transform": "translateX(-50%) scaleY(0.19387755102040813)",
            }
          }
        />
        <div
          className="MuiSlider-thumbWrapper MuiSlider-vertical"
          style={
            Object {
              "transform": "translateY(80.61224489795919%)",
            }
          }
        >
          <button
            aria-orientation="vertical"
            aria-valuemax={100}
            aria-valuemin={0}
            aria-valuenow={19.387755102040813}
            className="MuiButtonBase-root MuiSlider-thumb MuiSlider-vertical"
            onBlur={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            onTouchStartCapture={[Function]}
            role="slider"
            tabIndex={0}
            type="button"
          />
        </div>
        <div
          className="MuiSlider-track MuiSlider-trackAfter MuiSlider-vertical"
          style={
            Object {
              "transform": "translateX(-50%) scaleY(0.8061224489795918)",
            }
          }
        />
      </div>
    </div>
  </div>
  <canvas
    className="sc-gzVnrw fgSTpW"
    height={8000}
    id="animationPaneCanvas"
    width={16000}
  />
</div>
`;

exports[`simple component tests Channel component 1`] = `
<div
  className="WaveformCanvases sc-dnqmqq kKGkmx"
  offset={0}
/>
`;
